### PR TITLE
Fix for issue #18: values of metrics are constant - getStats() is called only once in init

### PR DESCRIPTION
**Pull request by obourdon**
Has been successfully tested in my environment

### CODE REVIEW

**Comment by obourdon** [comment]
Please note however that this PR needs to be modified so as to be compatible with [PR#23](https://github.com/intelsdi-x/snap-plugin-collector-cpu/pull/23) because the added getStats in cpu/cpu.go:157 requires one more argument p.proc_path as its 1st arg.

File: cpu/cpu.go [hunk 1]
```diff
@@ -146,12 +146,20 @@ func (p *Plugin) GetMetricTypes(cfg plugin.ConfigType) ([]plugin.MetricType, err
 // It returns error in case retrieval was not successful
 func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.MetricType, error) {
 	metrics := []plugin.MetricType{}
-	for _, metricType := range metricTypes {
-		if !p.initialized {
-			if err := p.init(metricType.Config().Table()); err != nil {
-				return nil, err
-			}
+	if len(metricTypes) == 0 {
```
Comment (marcin-krolik) [inline cpu/cpu.go]: This check is not needed. It's safe to assume `len(metricTypes) > 0`

File: cpu/cpu.go [hunk 2]
```diff
@@ -146,12 +146,20 @@ func (p *Plugin) GetMetricTypes(cfg plugin.ConfigType) ([]plugin.MetricType, err
...
+		return nil, nil
+	}
+	if !p.initialized {
+		if err := p.init(metricTypes[0].Config().Table()); err != nil {
+			return nil, err
 		}
+	} else {
```
Comment (marcin-krolik) [inline cpu/cpu.go]: `else` can be skipped here

**Comment by marcin-krolik** [comment]
@obourdon Since plugin is using configuration in `CollectMetrics()` what do you think about adding proper implementation of `GetConfigPolicy()`?

File: cpu/cpu.go (hunk 2 above)
Comment (obourdon) [inline cpu/cpu.go]: Not sure I understand your point here, the purpose of the else is the very core of the fix for this issue: getStats not being called in subsequent calls of metric gathering. Please advise

Comment (marcin-krolik) [inline cpu/cpu.go]: @obourdon Sorry for not being precise. I forgot to add that if plugin is not initiliazed, `p.init` is called. Inside `p.init` apart from reading configuration, there is a [call](https://github.com/obourdon/snap-plugin-collector-cpu/blob/e59905586ee357391e1eecab6f7e1f0178038b50/cpu/cpu.go#L226) to `getStats`. I think `p.init` should not call `getStats`, should only initialize state of the plugin, so later on `getStats` can be called safely.

I suggest to remove `getStats` from `p.init`, and then above block can be rewritten to

```go
if !p.initialized {
	if err := p.init(metricTypes[0].Config().Table()); err != nil {
		return nil, err
	}
}

if err := getStats(p.stats, p.prevMetricsSum, p.cpuMetricsNumber, p.snapMetricsNames, p.procStatMetricsNames); err != nil {
	return nil, err
 }
```

Comment (obourdon) [inline cpu/cpu.go]: @marcin-krolik ok, I now understand what you meant and I agree with you. I did not like the fact that init was implicitly calling getStats either. However, as I did not write this part initially I did not want to change it furthermore as I might not have all background/expertise to infer all potential impacts. Just submitted a new version integrating this change. Thanks for your comments

Comment (marcin-krolik) [inline cpu/cpu.go]: @obourdon I will fix it later then, when your PR will be merged

**Comment by katarzyna-z** [comment]
@obourdon Would you please rebase your code?

**Comment by obourdon** [comment]
@katarzyna-z rebase has been done and tests are now working

**Comment by katarzyna-z** [comment]
LGTM

**Pull request merged**